rfcs/rfc-create-ui-architecture.md

```diff
@@ -0,0 +1,146 @@
+
+(*) Although default/overarching rules, with specific exceptions for some filing types, initially sound like a good idea, their use should be minimized or heavily scrutinized, as they can lead to complicated logic -- which will lead to complicated or hard-to-understand (and therefore, hard to implement, debug and maintain) code.
+
+## Principles
+
+- start small and add to it
+- keep a sane architecture as we build upon it
+- keep it general so that it can be extended easily
+- refactor resources/config object as needed
+- minimize code refactoring
+- create unit tests to ensure resource object and components continue to function through code changes
+- parameterize unit tests to minimize duplication
+- create and use common or [shared components](https://github.com/bcgov/bcrs-shared-components) as opportunities arise
+
+## Implementation
+
+### Option 1
+
+- implement/enhance a component once with all possible permutations to handle all different filings
+- difficult or impossible to build correctly (we can't know everything up front)
+- not possible to test end-to-end during development
+- cannot release (deploy) new filing types one by one
+
+### Option 2
+
+- add filings types one by one
+- enhance components as we go, adding in per-type handling where needed
+- this is likely easier and more flexible as we only update what we need, when we need it (ie, Agile)
+- this is probably less work, even though we'll touch components often
+- can release new filing types as they are implemented
+
+### Recommendation
+
+Follow option 2.
+
+## Grooming Considerations
+
+1. should use a story per filing type (instead of a story per component)
+2. within stories, add tasks for different application steps/pages
+3. first story for a new filing type (other than IA) needs extra time to design and implement the resource/config object
+4. first story also needs extra time to create initial (parameterized) unit tests
+5. subsequent stories will need time to update the resources/config object
+6. subsequent stories will need time to update code where functionality differs between filing types
+7. subsequent stories will need time to update unit tests
+8. are there any dependencies? do we know what we need to, to begin?
+9. can we update a schema(s) so that the API can validate the "business logic" captured in the data? which other non-schema API validations will be needed?
+
+## Complications/Outstanding Questions
+
+1. we should investigate existing mechanisms/libraries to manage the resources object
+   - [vue-config](https://github.com/alex-oleshkevich/vue-config) may have code we can use to access a "config"
+   - see what we did for resources in Filings UI and maybe build upon it
+2. should we implement i18n for internationalization (ie, support for English/French/FN) at the same time?
+   - see [vue-i18n](https://kazupon.github.io/vue-i18n/)
+   - this is probably different enough that there is little value in doing it as the same time
```

Comment:
I strongly disagree with this statement.
It is not like Babel, which picks strings from the app and provides a replacement.
I suggested we look to leverage at a minimum the approach of wrapping and key.
eg ```<p>{{ $t('message.hello', { msg: 'hello' }) }}</p>```

Reply:
As discussed during grooming, we should:
- create our own access functions
- use a similar format/syntax to i18n (which some devs are familiar with) to pull our data (per Thor's example above)
- can "wrap" i18n inside our access functions
- could leverage vue-config inside our access functions

Reply:
Updated.